TestParallelDial is not stable

## Build
https://github.com/networkservicemesh/sdk/actions/runs/163461339

## Logs
```
=== FAIL: pkg/networkservice/common/connect TestParallelDial (0.98s)
##[error]    leaks.go:78: found unexpected goroutines:
        [Goroutine 59 in state select, with google.golang.org/grpc.(*ccBalancerWrapper).watcher on top of the stack:
        goroutine 59 [select]:
        google.golang.org/grpc.(*ccBalancerWrapper).watcher(0xc000050f80)
        	C:/Users/runneradmin/go/pkg/mod/google.golang.org/grpc@v1.29.1/balancer_conn_wrappers.go:69 +0x1a9
        created by google.golang.org/grpc.newCCBalancerWrapper
        	C:/Users/runneradmin/go/pkg/mod/google.golang.org/grpc@v1.29.1/balancer_conn_wrappers.go:60 +0x2fb
        
         Goroutine 60 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
        goroutine 60 [IO wait]:
```

--- a/pkg/networkservice/common/connect/server_test.go
+++ b/pkg/networkservice/common/connect/server_test.go
@@ -14,7 +14,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package connect
+package connect_test
 
 import (
 	"context"
@@ -24,28 +24,31 @@ import (
 	"testing"
 	"time"
 
-	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
-
-	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
-	"github.com/networkservicemesh/sdk/pkg/networkservice/common/setextracontext"
-
+	"github.com/google/uuid"
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"google.golang.org/grpc"
 
-	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/client"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/chains/endpoint"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/authorize"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/connect"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/refresh"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/setextracontext"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/updatepath"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/common/updatetoken"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/adapters"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/inject/injectpeer"
+	"github.com/networkservicemesh/sdk/pkg/tools/clienturlctx"
+	"github.com/networkservicemesh/sdk/pkg/tools/sandbox"
+	"github.com/networkservicemesh/sdk/pkg/tools/token"
 
 	"github.com/stretchr/testify/require"
 	"go.uber.org/goleak"
-	"google.golang.org/grpc/credentials"
 )
 
 const (
 	timeout = 10 * time.Second
 )
-
-func TokenGenerator(peerAuthInfo credentials.AuthInfo) (token string, expireTime time.Time, err error) {
-	return "TestToken", time.Date(3000, 1, 1, 1, 1, 1, 1, time.UTC), nil
-}
 
 type nseTest struct {
 	ctx      context.Context
@@ -62,33 +65,33 @@ func (nseT *nseTest) Stop() {
 	<-nseT.errCh
 }
 
-type endpointImpl struct {
-	networkservice.NetworkServiceServer
-}
-
-func (e *endpointImpl) MonitorConnections(selector *networkservice.MonitorScopeSelector, server networkservice.MonitorConnection_MonitorConnectionsServer) error {
-	// Just dummy wait here
-	<-server.Context().Done()
-	return nil
-}
-
-func (e *endpointImpl) Register(s *grpc.Server) {
-	networkservice.RegisterNetworkServiceServer(s, e.NetworkServiceServer)
-	networkservice.RegisterMonitorConnectionServer(s, e)
-}
-
 func (nseT *nseTest) Setup() {
 	nseT.ctx, nseT.cancel = context.WithTimeout(context.Background(), 50*time.Second)
 	nseT.listenOn = &url.URL{Scheme: "tcp", Host: "127.0.0.1:0"}
-	nseT.nse = &endpointImpl{
-		NetworkServiceServer: setextracontext.NewServer(map[string]string{"ok": "all is ok"}),
-	}
+	nseT.nse = endpoint.NewServer(nseT.ctx, "testServer", authorize.NewServer(),
+		sandbox.GenerateTestToken,
+		setextracontext.NewServer(map[string]string{"ok": "all is ok"}))
 
 	nseT.errCh = endpoint.Serve(nseT.ctx, nseT.listenOn, nseT.nse)
 }
 
 func (nseT *nseTest) newNSEContext(ctx context.Context) context.Context {
 	return clienturlctx.WithClientURL(ctx, &url.URL{Scheme: "tcp", Host: nseT.listenOn.Host})
+}
+
+func newClient(ctx context.Context, generatorFunc token.GeneratorFunc) networkservice.NetworkServiceClient {
+	return chain.NewNetworkServiceClient(
+		append([]networkservice.NetworkServiceClient{
+			authorize.NewClient(),
+			updatepath.NewClient("nsc-" + uuid.New().String()),
+			refresh.NewClient(ctx),
+		},
+			injectpeer.NewClient(),
+			updatetoken.NewClient(generatorFunc),
+			adapters.NewServerToClient(connect.NewServer(ctx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
+				return networkservice.NewNetworkServiceClient(cc)
+			}, grpc.WithBlock(), grpc.WithInsecure())),
+		)...)
 }
 
 func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
@@ -102,15 +105,12 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 
 	t.Run("Check Request", func(t *testing.T) {
 		require.NotPanics(t, func() {
-			serverCtx, serverCancel := context.WithCancel(context.Background())
-			defer serverCancel()
-			s := NewServer(serverCtx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-				return networkservice.NewNetworkServiceClient(cc)
-			}, grpc.WithInsecure())
 			clientURLCtx, clientCancel := context.WithTimeout(context.Background(), timeout)
 			defer clientCancel()
 			clientURLCtx = nseT.newNSEContext(clientURLCtx)
-			conn, err := s.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
+
+			c := newClient(clientURLCtx, sandbox.GenerateTestToken)
+			conn, err := c.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
 				Connection: &networkservice.Connection{
 					Id: "1",
 				},
@@ -118,7 +118,7 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 			require.Nil(t, err)
 			require.NotNil(t, conn)
 			require.Equal(t, "all is ok", conn.GetContext().GetExtraContext()["ok"])
-			_, err = s.Close(clientURLCtx, &networkservice.Connection{
+			_, err = c.Close(clientURLCtx, &networkservice.Connection{
 				Id: conn.Id,
 			})
 			require.Nil(t, err)
@@ -126,15 +126,12 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 	})
 	t.Run("Close Id", func(t *testing.T) {
 		require.NotPanics(t, func() {
-			serverCtx, serverCancel := context.WithCancel(context.Background())
-			defer serverCancel()
-			s := NewServer(serverCtx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-				return networkservice.NewNetworkServiceClient(cc)
-			}, grpc.WithInsecure())
 			clientURLCtx, clientCancel := context.WithTimeout(context.Background(), timeout)
 			defer clientCancel()
 			clientURLCtx = nseT.newNSEContext(clientURLCtx)
-			conn, err := s.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
+
+			c := newClient(clientURLCtx, sandbox.GenerateTestToken)
+			conn, err := c.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
 				Connection: &networkservice.Connection{
 					Id: "1",
 				},
@@ -144,7 +141,7 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 			require.Equal(t, "all is ok", conn.GetContext().GetExtraContext()["ok"])
 
 			// Do not pass clientURL
-			_, err = s.Close(context.Background(), &networkservice.Connection{
+			_, err = c.Close(context.Background(), &networkservice.Connection{
 				Id: "1",
 			})
 			require.Nil(t, err)
@@ -152,13 +149,11 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 	})
 	t.Run("Check no clientURL", func(t *testing.T) {
 		require.NotPanics(t, func() {
-			serverCtx, serverCancel := context.WithCancel(context.Background())
-			defer serverCancel()
-			s := NewServer(serverCtx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-				return networkservice.NewNetworkServiceClient(cc)
-			}, grpc.WithInsecure())
+			clientURLCtx, clientCancel := context.WithTimeout(context.Background(), timeout)
+			defer clientCancel()
 
-			conn, err := s.Request(context.Background(), &networkservice.NetworkServiceRequest{
+			c := newClient(clientURLCtx, sandbox.GenerateTestToken)
+			conn, err := c.Request(context.Background(), &networkservice.NetworkServiceRequest{
 				Connection: &networkservice.Connection{
 					Id: "1",
 				},
@@ -169,15 +164,12 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 	})
 	t.Run("Request without client URL", func(t *testing.T) {
 		require.NotPanics(t, func() {
-			serverCtx, serverCancel := context.WithCancel(context.Background())
-			defer serverCancel()
-			s := NewServer(serverCtx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-				return networkservice.NewNetworkServiceClient(cc)
-			}, grpc.WithInsecure())
 			clientURLCtx, clientCancel := context.WithTimeout(context.Background(), timeout)
 			defer clientCancel()
 			clientURLCtx = nseT.newNSEContext(clientURLCtx)
-			conn, err := s.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
+
+			c := newClient(clientURLCtx, sandbox.GenerateTestToken)
+			conn, err := c.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
 				Connection: &networkservice.Connection{
 					Id: "1",
 				},
@@ -187,7 +179,7 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 			require.Equal(t, "all is ok", conn.GetContext().GetExtraContext()["ok"])
 
 			// Request again
-			conn, err = s.Request(context.Background(), &networkservice.NetworkServiceRequest{
+			conn, err = c.Request(context.Background(), &networkservice.NetworkServiceRequest{
 				Connection: &networkservice.Connection{
 					Id: "1",
 				},
@@ -197,7 +189,7 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 			require.Equal(t, "all is ok", conn.GetContext().GetExtraContext()["ok"])
 
 			// Do not pass clientURL
-			_, err = s.Close(context.Background(), &networkservice.Connection{
+			_, err = c.Close(context.Background(), &networkservice.Connection{
 				Id: "1",
 			})
 			require.Nil(t, err)
@@ -206,17 +198,12 @@ func TestConnectServerShouldNotPanicOnRequest(t *testing.T) {
 }
 
 func TestParallelDial(t *testing.T) {
-	t.Skip("https://github.com/networkservicemesh/sdk/issues/377")
 	defer goleak.VerifyNone(t, goleak.IgnoreCurrent())
 
 	nseT := &nseTest{}
 	nseT.Setup()
 	defer nseT.Stop()
 
-	serverCtx, serverCancel := context.WithCancel(context.Background())
-	defer serverCancel()
-
-	s := NewServer(serverCtx, client.NewClientFactory("nsc", nil, TokenGenerator), grpc.WithInsecure())
 	var wg sync.WaitGroup
 	for i := 0; i < 10; i++ {
 		wg.Add(1)
@@ -227,18 +214,16 @@ func TestParallelDial(t *testing.T) {
 		go func() {
 			defer wg.Done()
 			for k := 0; k < 10; k++ {
-				conn, err := s.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
+				c := newClient(clientURLCtx, sandbox.GenerateTestToken)
+				conn, err := c.Request(clientURLCtx, &networkservice.NetworkServiceRequest{
 					Connection: &networkservice.Connection{
 						Id: fmt.Sprintf("%d", j),
-						Path: &networkservice.Path{
-							PathSegments: []*networkservice.PathSegment{},
-						},
 					},
 				})
 				require.Nil(t, err)
 				require.NotNil(t, conn)
 
-				_, err = s.Close(clientURLCtx, conn)
+				_, err = c.Close(clientURLCtx, conn)
 
 				require.Nil(t, err)
 			}

